dialer/fastconnect.go

```diff
@@ -0,0 +1,198 @@
+package dialer
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"sort"
+	"sync"
+	"time"
+)
+
+type connectTimeProxyDialer struct {
+	ProxyDialer
+
+	connectTime time.Duration
+}
+
+type dialersByConnectTime []connectTimeProxyDialer
+
+func (d dialersByConnectTime) Len() int {
+	return len(d)
+}
+
+func (d dialersByConnectTime) Less(i, j int) bool {
+	return d[i].connectTime < d[j].connectTime
+}
+
+func (d dialersByConnectTime) Swap(i, j int) {
+	d[i], d[j] = d[j], d[i]
+}
+
+// fastConnectDialer stores the time it took to connect to each dialer and uses
+// that information to select the fastest dialer to use.
+type fastConnectDialer struct {
+	topDialer     protectedDialer
+	connected     dialersByConnectTime
+	connectedChan chan int
+	// Lock for the slice of dialers.
+	connectedLock sync.RWMutex
+
+	next func(*Options, Dialer) Dialer
+	opts *Options
+}
+
+func newFastConnectDialer(opts *Options, next func(opts *Options, existing Dialer) Dialer) *fastConnectDialer {
+	if opts.OnError == nil {
+		opts.OnError = func(error, bool) {}
+	}
+	if opts.OnSuccess == nil {
+		opts.OnSuccess = func(ProxyDialer) {}
+	}
+	return &fastConnectDialer{
+		connected:     make(dialersByConnectTime, 0),
+		connectedChan: make(chan int),
+		opts:          opts,
+		next:          next,
+	}
+}
+
+func (fcd *fastConnectDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	// Use the dialer with the lowest connect time, waiting on early dials for any
+	// connections at all.
+	td := fcd.topDialer.get()
+	if td == nil {
+		return nil, fmt.Errorf("no top dialer")
+	}
+
+	// Note that we don't currently check if the dialer supports
+	// the domain here.
+	conn, failedUpstream, err := td.DialContext(ctx, network, addr)
+	if err != nil {
+		hasSucceeding := len(fcd.connected) > 0
+		fcd.opts.OnError(err, hasSucceeding)
+		// Error connecting to the proxy or to the destination
+		if failedUpstream {
+			// Error connecting to the destination
+			log.Debugf("Error connecting to upstream destination %v: %v", addr, err)
+		} else {
+			// Error connecting to the proxy
+			log.Debugf("Error connecting to proxy %v: %v", td.Name(), err)
+		}
+		return nil, err
+	}
+	fcd.opts.OnSuccess(td)
+	return conn, err
+}
+
+func (fcd *fastConnectDialer) onConnected(pd ProxyDialer, connectTime time.Duration) {
+	log.Debugf("Connected to %v", pd.Name())
+	fcd.connectedLock.Lock()
+	defer fcd.connectedLock.Unlock()
+
+	fcd.connected = append(fcd.connected, connectTimeProxyDialer{
+		ProxyDialer: pd,
+		connectTime: connectTime,
+	})
+	sort.Sort(fcd.connected)
```

Comment:
just FYI, there's https://pkg.go.dev/slices#SortFunc if you want to avoid overriding swap/less methods and introducing new type

Reply:
I like it! Was poking around for something better but never found that.